File: flatten_binary_tree_to_list.py

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def in_order(root):
    if root is None:
        return
    in_order(root.left)
    print root.val
    in_order(root.right)

class Solution:
    def flatten_helper(self,root):
        if root is None:
            return None,None
        l_h, l_t = self.flatten_helper(root.left)
        r_h, r_t = self.flatten_helper(root.right)
        new_head = root
        if l_h:
            root.right = l_h
        new_tail = l_t if l_t else root
        if r_h:
            new_tail.right = r_h
        new_tail = r_t if r_t else new_tail
        new_head.left = None
        new_tail.left = None
        return new_head,new_tail

    # @param root, a tree node
    # @return nothing, do it in place
    def flatten(self, root):
        if root is None:
            return
        self.flatten_helper(root)

if __name__ == "__main__":
    n1 = TreeNode(1)
    n2 = TreeNode(2)
    n3 = TreeNode(3)
    n4 = TreeNode(4)
    n5 = TreeNode(5)
    n6 = TreeNode(6)
    n1.left = n2
    n1.right = n5
    n2.left = n3
    n2.right = n4
    n5.right = n6
    in_order(n1)
    Solution().flatten(n1)
    print "---"
    in_order(n1)

File: flatten_binary_tree_to_list.cpp
/*
Given a binary tree, flatten it to a linked list in-place.

For example,
Given
         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:
   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
*/

#include "header.h"

void flatten(TreeNode* root, TreeNode*& head, TreeNode*& tail){
    if(!root){
        head = tail = NULL;
        return;
    }
    TreeNode* l_head = NULL, *l_tail=NULL, *r_head=NULL, *r_tail=NULL;
    flatten(root->left, l_head, l_tail);
    flatten(root->right, r_head, r_tail);
    root->left = NULL;
    root->right = l_head?l_head:r_head;
    if(l_head) l_tail->right=r_head;
    head=root;
    tail=r_tail?r_tail:l_tail?l_tail:head;
    tail->right = NULL;
}

void flatten(TreeNode* root){
    TreeNode* head = NULL, *tail=NULL;
    flatten(root, head, tail);
}

int main(){
    TreeNode* root = create_tree("12534#6");
    flatten(root);
    TreeNode* curr = root;
    while(curr){
        printf("%d ", curr->val);
        curr = curr->right;
    }
    printf("\n");
    return 0;
}
